test/extensions/filters/http/ext_proc/ext_proc_grpc_fuzz.cc

```diff
@@ -0,0 +1,315 @@
+    auto& encoder = encoder_decoder.first;
+
+    const uint32_t num_chunks =
+        fdp->ConsumeIntegralInRange<uint32_t>(0, ExtProcFuzzMaxStreamChunks);
+    for (uint32_t i = 0; i < num_chunks; i++) {
+      // If proxy closes connection before body is fully sent it causes a
+      // crash. To address this, the external processor sets a flag to
+      // signal when it has generated an immediate response which will close
+      // the connection in the future. We check this flag, which is protected
+      // by a lock, before sending a chunk. If the flag is set, we don't attempt
+      // to send more data, regardless of whether or not the
+      // codec_client connection is still open. There are no locks protecting
+      // the codec_client connection and cannot trust that it's safe to send
+      // another chunk
+      fh->immediate_resp_lock_.lock();
+      if (fh->immediate_resp_sent_) {
+        ENVOY_LOG_MISC(trace, "Proxy closed connection, returning early");
+        fh->immediate_resp_lock_.unlock();
+        return response;
+      }
+      const uint32_t data_size = fdp->ConsumeIntegralInRange<uint32_t>(0, ExtProcFuzzMaxDataSize);
+      ENVOY_LOG_MISC(trace, "Sending chunk of {} bytes", data_size);
+      codec_client_->sendData(encoder, data_size, false);
+      fh->immediate_resp_lock_.unlock();
+    }
+
+    // See comment above
+    fh->immediate_resp_lock_.lock();
+    if (!fh->immediate_resp_sent_) {
+      ENVOY_LOG_MISC(trace, "Sending empty chunk to close stream");
+      Buffer::OwnedImpl empty_chunk;
+      codec_client_->sendData(encoder, empty_chunk, true);
+    }
+    fh->immediate_resp_lock_.unlock();
+    return response;
+  }
+
+  IntegrationStreamDecoderPtr randomDownstreamRequest(FuzzedDataProvider* fdp,
+                                                      ExtProcFuzzHelper* fh) {
+    // From the external processor's view each of these requests
+    // are handled the same way. They only differ in what the server should
+    // send back to the client.
+    // TODO(ikepolinsky): add random flag for sending trailers with a request
+    //   using HttpIntegration::sendTrailers()
+    switch (fdp->ConsumeEnum<HttpMethod>()) {
+    case HttpMethod::GET:
+      ENVOY_LOG_MISC(trace, "Sending GET request");
+      return sendDownstreamRequest(absl::nullopt);
+    case HttpMethod::POST:
+      if (fdp->ConsumeBool()) {
+        ENVOY_LOG_MISC(trace, "Sending POST request with body");
+        const uint32_t data_size = fdp->ConsumeIntegralInRange<uint32_t>(0, ExtProcFuzzMaxDataSize);
+        const std::string data = std::string(data_size, 'a');
+        return sendDownstreamRequestWithBody(data, absl::nullopt);
+      } else {
+        ENVOY_LOG_MISC(trace, "Sending POST request with chunked body");
+        return sendDownstreamRequestWithChunks(fdp, fh, absl::nullopt);
+      }
+    default:
+      ENVOY_LOG_MISC(error, "Unhandled HttpMethod");
+      exit(EXIT_FAILURE);
```

Comment:
You can probably just tmake this a `RELEASE_ASSERT` with the error message.

Reply:
Updated exit to RELEASE_ASSERT in `ext_proc_grpc_fuzz_helper.cc` too.